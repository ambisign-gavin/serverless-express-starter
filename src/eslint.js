// @flow
import execa from 'execa';
import fs from 'fs';
import { join } from 'path';

class Eslint {
    getPackages(): Array<string> {
        return [
            'eslint@^5.0.0',
            'babel-eslint@^10.0.0'
        ];
    }

    async init(projectPath: string): execa.ExecaChildProcess {
        return execa('node_modules/eslint/bin/eslint.js',[
            '--init'
        ], {
            cwd: projectPath,
            stdio: 'inherit'
        });
    }

    rejectParserConfig(projectPath: string) {
        const jsConfig = join(projectPath, '.eslintrc.js');
        if (fs.existsSync(jsConfig)) {
            let eslintrc: { default: Object, parser: string } = require(jsConfig);
            delete eslintrc['default'];
            eslintrc['parser'] = 'babel-eslint';
            const newEslintrc = 'module.exports = ' + JSON.stringify(eslintrc, null, 4) + '\n';
            fs.writeFileSync(jsConfig, newEslintrc);
        }

        const jsonConfig = join(projectPath, '.eslintrc.json');
        if (fs.existsSync(jsonConfig)) {
            let eslintrc = require(jsonConfig);
            eslintrc = {
                ...eslintrc,
                parser: 'babel-eslint'
            };
            const newEslintrc = JSON.stringify(eslintrc, null, 4) + '\n';
            fs.writeFileSync(jsonConfig, newEslintrc);
        }
    }
}

const eslint = new Eslint();

export default eslint;