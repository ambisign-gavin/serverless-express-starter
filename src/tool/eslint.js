// @flow
import execa from 'execa';
import fs from 'fs';
import { join } from 'path';
import YAML from 'yaml';

class Eslint {
    getInstallPackages(): Array<string> {
        return [
            'eslint@^5.0.0',
            'babel-eslint@^10.0.0'
        ];
    }

    init(projectPath: string): execa.ExecaChildProcess {
        return execa('node_modules/eslint/bin/eslint.js',[
            '--init'
        ], {
            cwd: projectPath,
            stdio: 'inherit'
        });
    }

    loadConfig(configPath: string): Object {
        let config = {};
        const jsConfig = join(configPath, '.eslintrc.js');
        if (fs.existsSync(jsConfig)) {
            config = require(jsConfig);
        }

        const jsonConfig = join(configPath, '.eslintrc.json');
        if (fs.existsSync(jsonConfig)) {
            config = require(jsonConfig);
        }

        const ymlConfig = join(configPath, '.eslintrc.yml');
        if (fs.existsSync(ymlConfig)) {
            const file = fs.readFileSync(ymlConfig, 'utf8');
            config = YAML.parse(file);
        }
        return config;
    }

    saveConfig(configPath: string, config: Object) {
        const jsConfig = join(configPath, '.eslintrc.js');
        if (fs.existsSync(jsConfig)) {
            const newEslintrc = 'module.exports = ' + JSON.stringify(config, null, 4) + '\n';
            fs.writeFileSync(jsConfig, newEslintrc);
        }

        const jsonConfig = join(configPath, '.eslintrc.json');
        if (fs.existsSync(jsonConfig)) {
            const newEslintrc = JSON.stringify(config, null, 4) + '\n';
            fs.writeFileSync(jsonConfig, newEslintrc);
        }

        const ymlConfig = join(configPath, '.eslintrc.yml');
        if (fs.existsSync(ymlConfig)) {
            const newEslintrc = YAML.stringify(config);
            fs.writeFileSync(ymlConfig, newEslintrc);
        }
    }

    fixScripts(projectPath: string): execa.ExecaChildProcess {
        return execa('node_modules/eslint/bin/eslint.js',[
            '--fix',
            './src/*.js'
        ], {
            cwd: projectPath,
            stdio: 'inherit'
        });
    }

}

const eslint = new Eslint();

export default eslint;