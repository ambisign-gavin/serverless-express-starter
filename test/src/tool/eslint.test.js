// @flow
import eslint from '../../../src/tool/eslint';
import execa from 'execa';
import { join } from 'path';
import fs from 'fs';
import YAML from 'yaml';

jest.mock('execa');
jest.mock('fs');

describe('Eslint', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('should get install packages correct', () => {
        expect(eslint.getInstallPackages()).toEqual([
            'eslint@^5.0.0',
            'babel-eslint@^10.0.0'
        ]);
    });

    it('should init eslint correct', () => {
        eslint.init('path');
        expect(execa.mock.calls.length).toEqual(1);
        expect(execa.mock.calls[0][0]).toEqual('node_modules/eslint/bin/eslint.js');
        expect(execa.mock.calls[0][1]).toEqual([
            '--init'
        ]);
        expect(execa.mock.calls[0][2]).toEqual({
            cwd: 'path',
            stdio: 'inherit'
        });
    });

    it('should injection parser to js config correct', () => {
        const path = join(process.cwd(), 'test/configs');
        fs.existsSync.mockImplementation((path) => {
            return /.*(js)$/.test(path);
        });
        eslint.rejectParserConfig(path);
        expect(fs.writeFileSync.mock.calls.length).toEqual(1);
        expect(fs.writeFileSync.mock.calls[0][0]).toEqual(join(path, '.eslintrc.js'));
        expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should injection parser to json config correct', () => {
        const path = join(process.cwd(), 'test/configs');
        fs.existsSync.mockImplementation((path) => {
            return /.*(json)$/.test(path);
        });
        eslint.rejectParserConfig(path);
        expect(fs.writeFileSync.mock.calls.length).toEqual(1);
        expect(fs.writeFileSync.mock.calls[0][0]).toEqual(join(path, '.eslintrc.json'));
        expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should injection parser to yaml config correct', () => {
        const path = join(process.cwd(), 'test/configs');
        fs.existsSync.mockImplementation((path) => {
            return /.*(yml)$/.test(path);
        });
        jest.spyOn(YAML, 'parse').mockReturnValue({
            env: {
                es6: true,
                node: true
            },
            extends: 'eslint:recommended',
            globals: {
                Atomics: 'readonly',
                SharedArrayBuffer: 'readonly'
            },
            parserOptions: {
                ecmaVersion: 2018,
                sourceType: 'module'
            },
            rules: {}
        });
        eslint.rejectParserConfig(path);
        expect(fs.writeFileSync.mock.calls.length).toEqual(1);
        expect(fs.writeFileSync.mock.calls[0][0]).toEqual(join(path, '.eslintrc.yml'));
        expect(fs.writeFileSync.mock.calls[0][1]).toMatchSnapshot();
    });

    it('should fix scripts correct', () => {
        eslint.fixScripts('path');
        expect(execa.mock.calls.length).toEqual(1);
        expect(execa.mock.calls[0][0]).toEqual('node_modules/eslint/bin/eslint.js');
        expect(execa.mock.calls[0][1]).toEqual([
            '--fix',
            './src/*.js'
        ]);
        expect(execa.mock.calls[0][2]).toEqual({
            cwd: 'path',
            stdio: 'inherit'
        });
    });

});
