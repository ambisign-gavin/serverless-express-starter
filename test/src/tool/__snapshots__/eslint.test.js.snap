// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Eslint should injection parser to js config correct 1`] = `
"module.exports = {
    \\"env\\": {
        \\"es6\\": true,
        \\"node\\": true
    },
    \\"extends\\": \\"eslint:recommended\\",
    \\"globals\\": {
        \\"Atomics\\": \\"readonly\\",
        \\"SharedArrayBuffer\\": \\"readonly\\"
    },
    \\"parserOptions\\": {
        \\"ecmaVersion\\": 2018,
        \\"sourceType\\": \\"module\\"
    },
    \\"rules\\": {},
    \\"parser\\": \\"babel-eslint\\"
}
"
`;

exports[`Eslint should injection parser to json config correct 1`] = `
"{
    \\"env\\": {
        \\"es6\\": true,
        \\"node\\": true
    },
    \\"extends\\": \\"eslint:recommended\\",
    \\"globals\\": {
        \\"Atomics\\": \\"readonly\\",
        \\"SharedArrayBuffer\\": \\"readonly\\"
    },
    \\"parserOptions\\": {
        \\"ecmaVersion\\": 2018,
        \\"sourceType\\": \\"module\\"
    },
    \\"rules\\": {},
    \\"parser\\": \\"babel-eslint\\"
}
"
`;

exports[`Eslint should injection parser to yaml config correct 1`] = `
"env:
  es6: true
  node: true
extends: eslint:recommended
globals:
  Atomics: readonly
  SharedArrayBuffer: readonly
parserOptions:
  ecmaVersion: 2018
  sourceType: module
rules: {}
parser: babel-eslint
"
`;
